app/code/Magento/Backend/Controller/Adminhtml/Index/GlobalSearch.php

```diff
@@ -61,34 +75,64 @@ public function execute()
                         'Please make sure that all global admin search modules are installed and activated.'
                     ),
                 ];
-            } else {
-                $start = $this->getRequest()->getParam('start', 1);
-                $limit = $this->getRequest()->getParam('limit', 10);
-                $query = $this->getRequest()->getParam('query', '');
-                foreach ($this->_searchModules as $searchConfig) {
-                    if ($searchConfig['acl'] && !$this->_authorization->isAllowed($searchConfig['acl'])) {
-                        continue;
-                    }
-
-                    $className = $searchConfig['class'];
-                    if (empty($className)) {
-                        continue;
-                    }
-                    $searchInstance = $this->_objectManager->create($className);
-                    $results = $searchInstance->setStart(
-                        $start
-                    )->setLimit(
-                        $limit
-                    )->setQuery(
-                        $query
-                    )->load()->getResults();
-                    $items = array_merge_recursive($items, $results);
-                }
             }
         }
 
         /** @var \Magento\Framework\Controller\Result\Json $resultJson */
         $resultJson = $this->resultJsonFactory->create();
         return $resultJson->setData($items);
     }
+
+    /**
+     * @return array
+     */
+    protected function getPreviewItems()
+    {
+        $result = [];
+        $query = $this->getRequest()->getParam('query', '');
+        foreach ($this->previewModules as $previewConfig) {
+            if ($previewConfig['acl'] && !$this->_authorization->isAllowed($previewConfig['acl'])) {
+                continue;
+            }
+            if (!$previewConfig['url'] || !$previewConfig['text']) {
+                continue;
+            }
+            $result[] = [
+                'url' => $this->getUrl($previewConfig['url']).'?search='.$query,
+                'name' => __($previewConfig['text'], $query)
+            ];
+        }
+        return $result;
+    }
+
+    /**
+     * @return array
+     */
+    protected function getSearchItems()
+    {
+        $items = [];
+        $start = $this->getRequest()->getParam('start', 1);
+        $limit = $this->getRequest()->getParam('limit', 10);
+        $query = $this->getRequest()->getParam('query', '');
+        foreach ($this->_searchModules as $searchConfig) {
+            if ($searchConfig['acl'] && !$this->_authorization->isAllowed($searchConfig['acl'])) {
+                continue;
+            }
+
+            $className = $searchConfig['class'];
+            if (empty($className)) {
+                continue;
+            }
+            $searchInstance = $this->_objectManager->create($className);
```

Comment:
why do we use ObjectManager here? 
the dedicated factory should be used instead. Btw, this factory should check the type of needed object . All the search classes should implement common interface/provide the same contract